src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c

```diff
@@ -620,39 +617,28 @@ event_callback(JNIEnv *env, EventInfo *evinfo)
     currentException = JNI_FUNC_PTR(env,ExceptionOccurred)(env);
     JNI_FUNC_PTR(env,ExceptionClear)(env);
 
-    /* See if a garbage collection finish event happened earlier.
-     *
-     * Note: The "if" is an optimization to avoid entering the lock on every
-     *       event; garbageCollected may be zapped before we enter
-     *       the lock but then this just becomes one big no-op.
-     */
-    if ( garbageCollected > 0 ) {
-        struct bag *unloadedSignatures = NULL;
+    /* check and process class unloading events */
+    struct bag *unloadedSignatures = NULL;
+
+    /* We also need to simulate the class unload events. */
+    debugMonitorEnter(handlerLock);
+
+    /* Analyze which class unloads occurred */
+    unloadedSignatures = classTrack_processUnloads(env);
+
+    debugMonitorExit(handlerLock);
 
+    /* Generate the synthetic class unload events and/or just cleanup.  */
+    if ( unloadedSignatures != NULL ) {
         /* We want to compact the hash table of all
          * objects sent to the front end by removing objects that have
          * been collected.
          */
         commonRef_compact();
```

Comment:
If so, should `cbGarbageCollectionFinish()` be a better place for the call?

Reply:
I'm not sure. I guess this would resolve any delay in doing the compact. You wouldn't need to wait for some other event to trigger it. On the other hand, it's not that important to do this in a timely manner, and it seems safer to keep it working the way it currently does. Putting it directly in `cbGarbageCollectionFinish` skips some locking and other checks (see BEGIN_CALLBACK and END_CALLBACK). If you are confident that putting it in `cbGarbageCollectionFinish` won't cause any problem, then I'm ok with it.